### PR TITLE
fix forward diff for Bayesian optimization

### PR DESCRIPTION
**Pull request by jbrea**
Hi @maximerischard 

With some of the newer changes I couldn't use `ForwardDiff` anymore in the acquisition function of `BayesianOptimization`. The fix required the right `eltype` in some functions ~but also going back to not use `BLAS.syrk!` in `_predict`~. Is this fine for you?

### CODE REVIEW

File: src/GPE.jl [hunk 1]
```diff
@@ -351,13 +351,17 @@ function _predict(gp::GPE, x::AbstractMatrix)
...
     Sigma_raw = cov(gp.kernel, x, x, priordata)
-    # Sigma_raw = Sigma_raw - Lck'Lck
-    LinearAlgebra.BLAS.syrk!('U', 'T', -1.0, Lck, 1.0, Sigma_raw)
-    LinearAlgebra.copytri!(Sigma_raw, 'U')
+    subtract_Lck!(eltype(x), Sigma_raw, Lck)
     # Add jitter to get stable covariance
     m, chol = make_posdef!(Sigma_raw)
     return mu, PDMat(m, chol)
 end
+@inline function subtract_Lck!(::Any, Sigma_raw, Lck)
```
Comment (maximerischard) [inline src/GPE.jl]: I would suggest inverting the type signatures of the two `subtract_Lck!`. `LinearAlgebra.BLAS.syrk!` only works for `AbstractMatrix{Float64}`. Your implementation of `subtract_Lck!(::Type{<:Dual}, Sigma_raw, Lck)` should be the generic one.

I'm not sure I see the need to have a first argument be the element type itself. I would just define `subtract_Lck!(Sigma_raw::AbstractMatrix, Lck::AbstractMatrix)` and `subtract_Lck!(Sigma_raw::AbstractMatrix{Float64}, Lck::AbstractMatrix{Float64})`.

File: src/kernels/distance.jl [hunk 2]
```diff
@@ -62,7 +62,13 @@ end
 
 # Euclidean
 @inline distij(dist::Euclidean,X::AbstractMatrix,i::Int,j::Int,dim::Int)=√_SqEuclidean_ij(X,i,j,dim)
-@inline distij(dist::Euclidean,X1::AbstractMatrix,X2::AbstractMatrix,i::Int,j::Int,dim::Int)=√_SqEuclidean_ij(X1,X2,i,j,dim)
+@inline function distij(dist::Euclidean,X1::AbstractMatrix,X2::AbstractMatrix,i::Int,j::Int,dim::Int)
+    if X1 === X2 && i == j
+        eltype(X2)(0)
```
Comment (maximerischard) [inline src/kernels/distance.jl]: would `zero(eltype(X2))` make more sense?

File: src/kernels/fixed_kernel.jl [hunk 3]
```diff
@@ -1,6 +1,6 @@
 struct FixedKernel{K<:Kernel, NFREE} <: Kernel
     kernel::K
-    free::SVector{NFREE, Int64}
+    free::SVector{NFREE, Int}
```
Comment (maximerischard) [inline src/kernels/fixed_kernel.jl]: My reasoning was that having a generic type in a `struct` will slow things down. I haven't benchmarked it though…

File: src/kernels/kernels.jl [hunk 4]
```diff
@@ -30,7 +30,7 @@ each column is an observation.
 """
 function cov(k::Kernel, X₁::AbstractMatrix, X₂::AbstractMatrix, kerneldata::KernelData=EmptyData())
     n1, n2 = size(X₁, 2), size(X₂, 2)
-    cK = Array{eltype(X₁)}(undef, n1, n2)
+    cK = Array{eltype(X₂)}(undef, n1, n2)
```
Comment (maximerischard) [inline src/kernels/kernels.jl]: I'm guessing this is more useful for your usecase, but it doesn't quite generalise. A better solution might be to compute a pilot value `cov_ij(k, X₁, X₂, 42, 43, dim)` and use its type to build cK. Or maybe just `promote(X₁[1,1], X₂[1,1])` or `promote_type(typeof(X₁), typeof(X₂))`?

File: test/kernels.jl [hunk 5]
```diff
@@ -129,6 +130,15 @@ function testkernel(kern::Kernel)
         update_target_and_dtarget!(gp)
         @test gp.dmll ≈ gp_empty.dmll rtol=1e-6 atol=1e-6
     end
+    
+    @testset "predict gradient" begin
```
Comment (maximerischard) [inline test/kernels.jl]: That's great to add a test for this. Thank you.

File: src/kernels/fixed_kernel.jl (hunk 3 above)
Comment (jbrea) [reply src/kernels/fixed_kernel.jl]: `Int` is not generic, if I understand [this](https://docs.julialang.org/en/v1/manual/integers-and-floating-point-numbers/#Integers-1) correctly.

File: src/GPE.jl (hunk 1 above)
Comment (jbrea) [reply src/GPE.jl]: Good point, thanks.

File: src/kernels/distance.jl (hunk 2 above)
Comment (jbrea) [reply src/kernels/distance.jl]: Yes.

File: src/kernels/kernels.jl (hunk 4 above)
Comment (jbrea) [reply src/kernels/kernels.jl]: Yes.

**Pull request merged**